File: app/src/main/java/com/slimshady/bottomsheetmodule/list/BottomSheetDynamicListMultiSelect.java
package com.slimshady.bottomsheetmodule.list;


import android.content.Context;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;

import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.slimshady.bottomsheetmodule.abstractions.IApplyButton;
import com.slimshady.bottomsheetmodule.abstractions.IDynamicList;
import com.slimshady.bottomsheetmodule.abstractions.ISearch;
import com.slimshady.bottomsheetmodule.adapter.AsyncSearchListAdapter;
import com.slimshady.bottomsheetmodule.adapter.listener.AdapterItemMultiSelectListener;
import com.slimshady.bottomsheetmodule.model.BaseBottomSheetRecyclerModel;
import com.slimshady.bottomsheetmodule.util.AdapterAction;
import com.slimshady.bottomsheetmodule.util.searchWatcher.SearchWatcher;

import java.util.ArrayList;

public class BottomSheetDynamicListMultiSelect<T extends BaseBottomSheetRecyclerModel, V extends ViewBinding> extends BaseBottomSheetRecyclerView<V> implements IDynamicList, IApplyButton, ISearch {

    private boolean isSearchEnable;
    private String searchHint;
    private ArrayList<T> items;
    private AdapterItemMultiSelectListener<T> adapterItemMultiSelectListener;


    private AsyncSearchListAdapter<T> asyncSearchListAdapter;

    private ConstraintLayout.LayoutParams buttonLayoutParams;
    private int collapsedMargin; //Button margin in collapsed state
    private int buttonHeight;

    private ArrayList<T> filteredListBaseSearchDbModel;

    public BottomSheetDynamicListMultiSelect(
            V viewBinding,
            Context context,
            int parentLayoutBottomSheetResId,
            RecyclerView.LayoutManager layoutManager,
            BaseBottomSheetRecyclerViewBuilder<V> bottomSheetRecyclerViewBuilder,
            boolean isSearchEnable,
            String searchHint,
            ArrayList<T> items,
            AdapterItemMultiSelectListener<T> adapterItemMultiSelectListener
    ) {

        super(viewBinding, context, parentLayoutBottomSheetResId, layoutManager, bottomSheetRecyclerViewBuilder);

        this.isSearchEnable = isSearchEnable;
        this.searchHint = searchHint;
        this.items = items;
        this.adapterItemMultiSelectListener = adapterItemMultiSelectListener;

        if (isSearchEnable) {
            searchView.setVisibility(View.VISIBLE);
            initSearchView(context, searchView, searchHint, closeBtn, searchIcon, searchEditText);
            closeButtonSearchListener();

        } else
            searchView.setVisibility(View.GONE);

        btnApply.setVisibility(View.VISIBLE);

        initListAdapter();
        initApplyButtonView();
        onClickApplyButton();
    }

    @Override
    public void initListAdapter() {
        asyncSearchListAdapter = new AsyncSearchListAdapter<T>(true);

        asyncSearchListAdapter.submitList(items);
        recyclerView.setAdapter(asyncSearchListAdapter);
        recyclerView.setItemAnimator(null);

    }

    @Override
    protected void initSearchView(Context context, SearchWatcher searchView, String searchHint, ImageView closeBtn, ImageView searchIcon, EditText searchEditText) {
        super.initSearchView(context, searchView, searchHint, closeBtn, searchIcon, searchEditText);

        searchView.addTextWatcher(s -> {
            if (s.trim().length() > 0) {
                filter(s, items);
            } else {
                asyncSearchListAdapter.submitList(items);
            }
        }, 400);
    }

    @Override
    public void initApplyButtonView() {
        buttonLayoutParams = (ConstraintLayout.LayoutParams) btnApply.getLayoutParams();

        //Calculate button margin from top
        buttonHeight = btnApply.getHeight(); //How tall is the button + experimental distance from bottom (Change based on your view)
        collapsedMargin = (int) (peekHeight - buttonHeight * 1.2); //Button margin in bottom sheet collapsed state
        buttonLayoutParams.topMargin = collapsedMargin;
        btnApply.setLayoutParams(buttonLayoutParams);
        recyclerView.setPadding(0,0,0, (int) (buttonHeight *2.2));

        Log.d("BottomSheetDynamic", "collapsedMargin: " + collapsedMargin + " expandedHeight: " + expandedHeight + " buttonHeight: "+ buttonHeight);

        initBottomSheetCallback(bottomSheetBehavior, btnApply, buttonLayoutParams, collapsedMargin, expandedHeight, buttonHeight);
    }

    @Override
    public void onClickApplyButton() {
        btnApply.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                adapterItemMultiSelectListener.onItemMultiSelect(asyncSearchListAdapter.getSelectedItems(), AdapterAction.SELECT);
                bottomSheetBehavior.setPeekHeight(0);
                bottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
            }
        });
    }

    private void filter(String query,
                        ArrayList<T> items) {
        filteredListBaseSearchDbModel = new ArrayList<>();

        for (int i = 0; i < items.size(); i++) {

            final String text = items.get(i).getName().toLowerCase();
            if (text.contains(query)) {

                filteredListBaseSearchDbModel.add(items.get(i));
            }
        }

        Log.d("BottomSheetSearch", "filteredList: " + filteredListBaseSearchDbModel.size());

        asyncSearchListAdapter = new AsyncSearchListAdapter<T>(true);

        recyclerView.setAdapter(asyncSearchListAdapter);
        asyncSearchListAdapter.submitList(filteredListBaseSearchDbModel);

    }

    @Override
    public void closeButtonSearchListener() {
        closeBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                searchView.clearFocus();
                searchView.setQuery("", false);
                bottomSheetBehavior.setPeekHeight(0);
                bottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
            }
        });
    }
}

File: app/src/main/java/com/slimshady/bottomsheetmodule/list/BottomSheetDynamicListMultiSelect.kt
package com.slimshady.bottomsheetmodule.list

import android.content.Context
import android.util.Log
import android.view.View

import com.slimshady.bottomsheetmodule.model.BaseBottomSheetRecyclerModel
import androidx.viewbinding.ViewBinding
import androidx.recyclerview.widget.RecyclerView
import com.slimshady.bottomsheetmodule.list.BaseBottomSheetRecyclerView.BaseBottomSheetRecyclerViewBuilder
import com.slimshady.bottomsheetmodule.adapter.listener.AdapterItemMultiSelectListener
import com.slimshady.bottomsheetmodule.list.BaseBottomSheetRecyclerView
import com.slimshady.bottomsheetmodule.abstractions.IDynamicList
import com.slimshady.bottomsheetmodule.abstractions.IApplyButton
import com.slimshady.bottomsheetmodule.abstractions.ISearch
import com.slimshady.bottomsheetmodule.adapter.AsyncSearchListAdapter
import androidx.constraintlayout.widget.ConstraintLayout
import com.slimshady.bottomsheetmodule.util.searchWatcher.SearchWatcher
import android.widget.EditText
import android.widget.ImageView
import com.slimshady.bottomsheetmodule.util.searchWatcher.Watcher
import com.slimshady.bottomsheetmodule.util.AdapterAction
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.slimshady.bottomsheetmodule.MainActivity
import java.util.*

class BottomSheetDynamicListMultiSelect<V : ViewBinding>(
    viewBinding: V,
    context: Context,
    parentLayoutBottomSheetResId: Int,
    layoutManager: RecyclerView.LayoutManager?,
    bottomSheetRecyclerViewBuilder: BaseBottomSheetRecyclerViewBuilder<V>?,
    private val isSearchEnable: Boolean,
    private val searchHint: String,
    private val items: ArrayList<BaseBottomSheetRecyclerModel>,
    private val adapterItemMultiSelectListener: MainActivity
) : BaseBottomSheetRecyclerView<V>(
    viewBinding,
    context,
    parentLayoutBottomSheetResId,
    layoutManager,
    bottomSheetRecyclerViewBuilder
), IDynamicList, IApplyButton, ISearch {
    private var asyncSearchListAdapter: AsyncSearchListAdapter<BaseBottomSheetRecyclerModel>? = null
    private var buttonLayoutParams: ConstraintLayout.LayoutParams? = null
    private var collapsedMargin //Button margin in collapsed state
            = 0
    private var buttonHeight = 0
    private var filteredListBaseSearchDbModel: ArrayList<BaseBottomSheetRecyclerModel>? = null
    override fun initListAdapter() {
        asyncSearchListAdapter = AsyncSearchListAdapter(true)
        asyncSearchListAdapter?.submitList(items)
        recyclerView.adapter = asyncSearchListAdapter
        recyclerView.itemAnimator = null
    }

    override fun initSearchView(
        context: Context,
        searchView: SearchWatcher,
        searchHint: String?,
        closeBtn: ImageView,
        searchIcon: ImageView,
        searchEditText: EditText
    ) {
        super.initSearchView(context, searchView, searchHint, closeBtn, searchIcon, searchEditText)
        searchView.addTextWatcher(object : Watcher {
            override fun onTextChange(s: String?) {
                if ((s?.trim { it <= ' ' }?.length ?: 0) > 0) {
                    if (s != null) {
                        filter(s, items)
                    }
                } else {
                    asyncSearchListAdapter?.submitList(items)
                }
            }
        }, 400)
    }

    override fun initApplyButtonView() {
        buttonLayoutParams = btnApply.layoutParams as ConstraintLayout.LayoutParams

        //Calculate button margin from top
        buttonHeight =
            btnApply.height //How tall is the button + experimental distance from bottom (Change based on your view)
        collapsedMargin =
            (peekHeight - buttonHeight * 1.2).toInt() //Button margin in bottom sheet collapsed state
        buttonLayoutParams?.topMargin = collapsedMargin
        btnApply.layoutParams = buttonLayoutParams
        recyclerView.setPadding(0, 0, 0, (buttonHeight * 2.2).toInt())
        initBottomSheetCallback(
            bottomSheetBehavior,
            btnApply,
            buttonLayoutParams,
            collapsedMargin,
            expandedHeight,
            buttonHeight
        )
    }

    override fun onClickApplyButton() {
        btnApply.setOnClickListener {
            adapterItemMultiSelectListener.onItemMultiSelect(
                asyncSearchListAdapter?.selectedItems,
                AdapterAction.SELECT
            )
            bottomSheetBehavior?.peekHeight = 0
            bottomSheetBehavior?.state = BottomSheetBehavior.STATE_COLLAPSED
        }
    }

    private fun filter(
        query: String,
        items: ArrayList<BaseBottomSheetRecyclerModel>
    ) {
        filteredListBaseSearchDbModel = ArrayList()
        for (i in items.indices) {
            val text = items[i].name?.lowercase(Locale.getDefault())
            if (text?.contains(query) == true) {
                filteredListBaseSearchDbModel?.add(items[i])
            }
        }
        Log.d("BottomSheetSearch", "filteredList: " + filteredListBaseSearchDbModel?.size)
        asyncSearchListAdapter = AsyncSearchListAdapter(true)
        recyclerView.adapter = asyncSearchListAdapter
        asyncSearchListAdapter?.submitList(filteredListBaseSearchDbModel)
    }

    override fun closeButtonSearchListener() {
        closeBtn.setOnClickListener {
            searchView.clearFocus()
            searchView.setQuery("", false)
            bottomSheetBehavior?.peekHeight = 0
            bottomSheetBehavior?.state = BottomSheetBehavior.STATE_COLLAPSED
        }
    }

    init {
        if (isSearchEnable) {
            searchView.visibility = View.VISIBLE
            initSearchView(context, searchView, searchHint, closeBtn, searchIcon, searchEditText)
            closeButtonSearchListener()
        } else searchView.visibility = View.GONE
        btnApply.visibility = View.VISIBLE
        initListAdapter()
        initApplyButtonView()
        onClickApplyButton()
    }
}